Add thread safety for Client Secret Credential

antkmsft opened the pull request:
Notes:
* No more mutex functions, `az_platform_atomic_compare_exchange()` instead.
* Moved `internal/az_platform_internal.h` to be `inc/az_platform.h`. Reason: customers are supposed to implement PAL layer with signatures from this header, so it should be public.
* Rearranged fields in credential structs for better alignment.
* We take lock to set, AND to get (just read and then quickly release). Without locking when get(), we cannot read the buffer reliably and assume it won't change just as we read it, as well we can't be sure that expiration and the token string do match.
* compare-exchange swaps pointers, so we have these two pointers that we are swapping, `&_az_token_credential_locked` and `&_az_token_credential_unlocked`. They are locally defined in the newly formed `az_credential_token.c` (Corresponds to TokenCredential in SDKs for other languages).
* I pulled `az_token_expired()` and `az_token_set()` to `az_credential_token.c`, as well as added `az_token_get()` (thread-safe), and `az_credential_token_init()`, which initializes lock pointer with `&_az_token_credential_unlocked`.


I recommend to read file changes (https://github.com/Azure/azure-sdk-for-c/pull/771/files) from the bottom, as it starts with platform, then `az_credential_token_private.h`, and moves into more details.

ahsonkhan commented on sdk/core/core/inc/az_credentials.h:
```diff
@@ -45,12 +46,21 @@ typedef struct
-    uint8_t token[_az_TOKEN_BUF_SIZE]; /*!< Base64-encoded token */
     int16_t token_length;
     int64_t expires_at_msec;
+    uint8_t token[_az_TOKEN_BUF_SIZE]; /*!< Base64-encoded token */
```
> Rearranged fields in credential structs for better alignment.

Can you please explain how this leads to better alignment?

antkmsft replied:
Ah, I should move int64_t before int16, my eyes didn't catch that initially.
Basically, on some platforms, compilers would align data to a word binary. It is usually int32, but can be any reasonable number (1, 2, 4, 8). If you put an array of 9 int8's first, and then an int32, it goes as `[int8][int8][int8][int8] | [int8][int8][int8][int8] | [int8][padding][padding][padding]|[int32]`

antkmsft replied:
Yes-yes, I acknowledged that right in my comment to Ahson above. I'll rearrange this.

antkmsft replied:
Done.

ahsonkhan replied:
Follow-up question: How are you figuring out how these struct fields are being laid out in memory to know that we are getting alignment? I understand the integer example, but what about array alignment, in C? Is that a pointer (word size) and int32 for the length?

antkmsft replied:
Array elements are always getting placed sequentially right next to each other so that `&array[0] + sizeof(array[0]) == &array[1]`. In this case specifically, we have an array of bytes, so it is going to be `_az_TOKEN_BUF_SIZE` number of bytes placed sequentially.

(`_az_TOKEN_BUF_SIZE` is 2048 bytes and I think it additionally might be a good practice to round byte array sizes to "binary round" numbers for this reason, although I am not completely sure enough to make that recommendation decisively - nothing bad would happen if it would be just 2000 bytes, but certainly it won't hurt if it is 2048. I personally like it for aesthetic reasons, and also with "binary round" numbers you can sometimes do tricks like bit masking, but again, not enough evidence to recommend it for all cases, other than personal preference)

JeffreyRichter replied:
The compiler will properly align fields (guaranteed) [unless you do something special like this: https://docs.microsoft.com/en-us/cpp/preprocessor/pack?view=vs-2019] - there is nothing we have to do. A byte will be on a byte boundary, a 16-bit field will be on a 16-bit boundary, a 32-bit field will be on a 32-bit boundary, and a 64-bit value will be on a 64-bit boundary. A pointer is either 32- or 64-bits depending on whether you compile for 32-bit or 64-bits. An array will be aligned as per the array type. So, in this case, a byte array is aligned on a byte boundary. For an array, it doesn't matter if its 2000 versus 2048 elements -- we should pick the number we want for the array; not a number that looks appealing in source code.

ahsonkhan commented on sdk/core/core/inc/az_credentials.h:
```diff
@@ -45,12 +46,21 @@ typedef struct
+{
+  struct
+  {
+    bool* volatile lock;
```
Do we have any atomic assignment operations in C? Is using `volatile` sufficient to implement a locking mechanism that is thread-safe?

antkmsft replied:
Oh do I wish it had!!! :) https://www.drdobbs.com/parallel/volatile-vs-volatile/212701484

antkmsft replied:
(resolving)

ahsonkhan commented on sdk/core/core/inc/az_platform.h:
```diff
@@ -0,0 +1,50 @@
+#include <_az_cfg_prefix.h>
+
+/**
+ * @brief az_platform_clock_msec gets platform clock in milliseconds.
```
Do not re-state the method name within the docs and make sure they are complete sentences.
```suggestion
 * @brief Gets the platform clock in milliseconds.
```

antkmsft replied:
Oh, we do it this way now? That's great, it is even better. I took a comment from another file as a template, we should fix it too.

ahsonkhan commented on sdk/core/core/inc/az_platform.h:
```diff
@@ -0,0 +1,50 @@
+
+/**
+ * @brief az_platform_clock_msec gets platform clock in milliseconds.
+ * The moment of time where clock starts is undefined, but if this function is getting called twice
+ * with one second interval, the difference between the values returned should be equal to 1000.
```
The `@brief` should be brief, ideally 1 sentence.  The rest should be in `@details`, `@note`, or `@remark`

```suggestion
 * @remark The moment of time where clock starts is undefined, but if this function is getting called twice
 * with one second interval, the difference between the values returned should be equal to 1000.
```

ahsonkhan commented on sdk/core/core/inc/az_platform.h:
```diff
@@ -0,0 +1,50 @@
+AZ_NODISCARD int64_t az_platform_clock_msec();
+
+/**
+ * @brief az_platform_sleep_msec tells platform to sleep for a given number of milliseconds.
```
```suggestion
 * @brief Tells the platform to sleep for a given number of milliseconds.
```

ahsonkhan commented on sdk/core/core/inc/az_platform.h:
```diff
@@ -0,0 +1,50 @@
+
+/**
+ * @brief az_platform_sleep_msec tells platform to sleep for a given number of milliseconds.
+ * @param milliseconds number of milliseconds to sleep.
```
```suggestion
 * @param milliseconds Number of milliseconds to sleep.
```

ahsonkhan commented on sdk/core/core/inc/az_platform.h:
```diff
@@ -0,0 +1,50 @@
+/**
+ * @brief az_platform_sleep_msec tells platform to sleep for a given number of milliseconds.
+ * @param milliseconds number of milliseconds to sleep.
+ * @remarks For passing \p milliseconds with negative values and 0, behavior is undefined.
```
nit: Reword
```suggestion
 * @remarks The behavior is undefined when \p milliseconds is a non-positive value (0 or less than 0).
```

ahsonkhan commented on sdk/core/core/inc/az_platform.h:
```diff
@@ -0,0 +1,50 @@
+void az_platform_sleep_msec(int32_t milliseconds);
+
+/**
+ * @brief az_platform_atomic_compare_exchange Conditionally exchanges values of two pointers in a
```
```suggestion
 * @brief Conditionally exchanges values of two pointers in a
```

ahsonkhan commented on sdk/core/core/inc/az_platform.h:
```diff
@@ -0,0 +1,50 @@
+ * operation otherwise. Returns `true` if the value was overwritten, `false` otherwise.
+ */
+AZ_NODISCARD bool az_platform_atomic_compare_exchange(
+    void* volatile* obj,
```
`volatile*` - What does this mean?

antkmsft replied:
So basically when you see a pointer, whatever is on the left side of `*` characterizes what it is pointing to. On the right - characterizes the pointer itself.
Remember this example?:
It is easier when you read pointers from right to left:
 `int const*` - non-constant pointer to constant int
`int* const` - constant pointer to non-constant int.
`int const* const` - constant pointer to constant int
`int*` - non-constant pointer to non-constant int.
... and so on, with pointers-to-pointers, it can take really monstrous shapes:
`int* const*** const* p` - p is a non-constant pointer to constant pointer to non-constant pointer to non-constant pointer to constant pointer to non-constant int.

A `const` is a modifier, and so is `volatile` (together, they are called "CV-modifiers"). Everywhere where you can use `const`, you can use `volatile`. You can use them both at the same time. Or just one. Or none of them.

And `void*` is "just" a pointer, to an unknown type. So, before you read or write to the value pointed by such pointer (`*p = 3`), you would need to cast it to a pointer of some type, other than `void`: *(int*)p = 3; *(double*)p = 2.0.

so, to read this, `void* volatile*` is a  (non-constant) pointer to a volatile pointer (to ???).

ahsonkhan commented on sdk/core/core/inc/az_platform.h:
```diff
@@ -0,0 +1,50 @@
+ * @return `true` if previous value of \p obj matches the \p expected.
+ *
+ * @remarks If `*obj` equals to \p expected, its value is being overwritten with \p desired, and no
+ * operation otherwise. Returns `true` if the value was overwritten, `false` otherwise.
```
> Returns `true` if the value was overwritten, `false` otherwise.

This belongs in `@return`

antkmsft replied:
I intended return to be short and factually/pedantically correct, while this section would better decipher would true/false means, in more simple words, but technically not as complete/accurate as @return section does.

ahsonkhan commented on sdk/platform/posix/src/az_posix.c:
```diff
@@ -21,25 +20,10 @@ void az_platform_sleep_msec(int32_t milliseconds)
-AZ_NODISCARD az_result az_platform_mtx_unlock(az_platform_mtx* mtx)
-{
-  return pthread_mutex_unlock(&mtx->_internal.mutex) == 0 ? AZ_OK : AZ_ERROR_MUTEX;
+  return __sync_bool_compare_and_swap(obj, expected, desired);
```
Will this work for all supported compilers?

antkmsft replied:
GCC and Clang for sure. One moment... It is a good question, I got a bit better idea.

antkmsft replied:
Updated, resolving.

ahsonkhan commented on sdk/platform/win32/src/az_win32.c:
```diff
@@ -1,34 +1,22 @@
-#include <az_platform_internal.h>
+#include <az_platform.h>
+
+#define WIN32_LEAN_AND_MEAN
+#define NOMINMAX
+#include <windows.h>
```
Can you please add a comment explaining these?

antkmsft replied:
These are pretty much common knowledge in the Windows programming world, these macros are needed to be defined before including `windows.h`. Yet despite being pretty widely used, there isn't a small MSDN page I could point to - the data is spread across MSDN here and there - https://docs.microsoft.com/en-us/search/?search=WIN32_LEAN_AND_MEAN . I'll see what I can do.

antkmsft replied:
Done, to an extent

ahsonkhan commented on sdk/core/core/src/az_credential_token_private.h:
```diff
@@ -0,0 +1,23 @@
+
+#include <_az_cfg_prefix.h>
+
+AZ_NODISCARD az_result _az_credential_token_init(_az_credential_token* out_credential_token);
```
nit: I understand these are private headers, but consider adding docs.

antkmsft replied:
At one point, let's do it all the way across our docs, review everything, do the renames for parameter prefies, replace `@return` with `@retval`s.

antkmsft replied:
(resolving, I updated #496 with the idea of documenting private headers)

ahsonkhan commented on sdk/core/core/inc/az_platform.h:
```diff
@@ -0,0 +1,50 @@
+ * @brief az_platform_atomic_compare_exchange Conditionally exchanges values of two pointers in a
+ * thread-safe manner.
+ *
+ * @param obj a pointer to a pointer that needs to be changed.
+ * @param expected an expected value of a value that \p obj is pointing to, prior to exchange.
+ * @param desired a value to assign to the value that is pointed by \p obj, if its value equals to
```
Capitalize. 

nit: Also, I am having difficulty understanding this a bit due to the repetition. Maybe there is a clearer way to explain the parameters.

antkmsft replied:
I'll try to see if I can rephrase this, maybe I'll borrow something from https://docs.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-interlockedcompareexchangepointer. Ah, documenting the code you just written feels a bit like playing chess against yourself :)

antkmsft replied:
I capitalized and tried to rephrase, resolving the conversation thread.

antkmsft commented on sdk/platform/win32/src/az_win32.c:
```diff
@@ -1,34 +1,22 @@
 {
-  LeaveCriticalSection(&mtx->_internal.cs);
-  return AZ_OK;
+  return InterlockedCompareExchangePointer(obj, expected, desired) == expected;
```
@JeffreyRichter, I'll need to update this. https://docs.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-interlockedcompareexchangepointer Says:
"On a 64-bit system, the parameters are 64 bits and must be aligned on 64-bit boundaries; otherwise, the function will behave unpredictably. On a 32-bit system, the parameters are 32 bits and must be aligned on 32-bit boundaries."
I think we need to store an array of 15 bytes, and then expect that whatever within this array ends up being aligned to a 64-bit boundary, that is our pointer. Because otherwise we need to require that the pointer is aligned, which means that the client needs to be aligned, which means a blob client, for example, can't be created on the stack.

JeffreyRichter replied:
Don't worry about this; the compiler WILL ensure that the pointer is aligned even if allocated on the stack.

antkmsft replied:
Currently, Linux 64 and only Linux 64 unit tests hang. It is hard to tell what's going on, because preconditions just sleep forever, and I don't know which one failed. But the symptoms are similar. Or maybe on Linux 64 + GCC, compiler does not do this alighnment, and fails into UB, that's another possibility. However `__sync_bool_compare_and_swap` do not mention requirement for alignment.

antkmsft replied:
I have a suspicion that GCC on Linux64 does not align the pointer and ends up shooting into access token buffer memory.

JeffreyRichter replied:
All compilers will align pointers. Plus, if a non-aligned value is accessed in a way that requires alignment, the system would not hang; it would generate an interrupt or exception instead. For example, on Windows, a EXCEPTION_DATATYPE_MISALIGNMENT exception is raised.

JeffreyRichter replied:
To debug what's happening, run the app under the debugger and put a breakpoint on the failed preconditon function.  Or, override the precondition callback by calling https://github.com/Azure/azure-sdk-for-c/blob/master/sdk/core/core/inc/az_precondition.h#L62 and have it to something else. For example, on Windows, you could call https://docs.microsoft.com/en-us/windows/win32/api/debugapi/nf-debugapi-debugbreak.  Or dump the stack somehow. It would be best to run in a debugger.

ahsonkhan replied:
> To debug what's happening, run the app under the debugger and put a breakpoint on the failed preconditon function.

FWIW, I couldn't get this approach to work in VS. Let me know if you are able to, @antkmsft

JeffreyRichter replied:
NOTE: You have the arguments in the wrong order: 
PVOID InterlockedCompareExchangePointer(
  PVOID volatile *Destination,
  PVOID          Exchange,
  PVOID          Comperand
);

requires "expected" to be the LAST argument and "desired" is the 2nd argument.

antkmsft replied:
All is corrected, and works. I am resolving this conversation thread.

antkmsft commented:
I addressed all the requests and suggestions for this PR, I fixed the alignment, and commented the usage of the GNU C language extension for atomic compare-exchange. 
Next:implement suggestions that Jeffrey has sent me over email, and fix the test failure that onlu occurs on Linux64: `az_span_is_content_equal(AZ_SPAN_FROM_STR("Bearer AccessToken"), header.value) test_az_credential_client_secret.c:200`

antkmsft commented:
@JeffreyRichter , this PR is ready to go. I don't have anything more to add to this PR, unless you have other requests. I fixed unit test and the bug we had, and I played with this code in debugger, both under Windows and Linux. If you review and everything looks good to you, please unblock this PR, and I'll check it in.

JeffreyRichter commented on sdk/core/az_core/src/az_credential_token.c:
```diff
@@ -0,0 +1,32 @@
+
+AZ_NODISCARD bool _az_token_expired(_az_token const* token)
+{
+  int64_t const expires_at_msec = token->_internal.expires_at_msec;
```
I think reading expires_at_msec should be done under the reader lock

antkmsft replied:
We get the token, with the locks. The token includes both the token string and the expiration time. Then you can read it safely to check expiration, or to include to headers. Read az_credential_client_secret.c in this PR, lines 55-71 is how we do it.
The token does not have a lock, token credential (_az_credential_token) does. _az_token_expired takes the _az_token, not the _az_credential_token.
This way, when you reuse-or-renew-and-use token in the `_az_credential_client_secret_apply()`, you get the token, with the lock, you check expiration, and if it is good, you apply it to the header. Otherwise, you send a request to obtain the new token (without locks), then you set it to the token credential (with locks), and use the token you just obtained and set, without locks to apply to the header.
It is just one read lock operation, instead of two. And you can be sure that as you read both the token string and the expiration, they two do match each other.

antkmsft replied:
Update: to discuss this, we talked over the phone, you are fine with the idea, and asked me to update comments, and I just pushed a commit with more comments. Once CI/CD completes, I am checking this PR in. Thank you! I am resolving this conversation thread.

JeffreyRichter commented on sdk/platform/noplatform/src/az_noplatform.c:
```diff
@@ -1,10 +1,21 @@
 void az_platform_sleep_msec(int32_t milliseconds) { (void)milliseconds; }
+
+AZ_NODISCARD bool az_platform_atomic_compare_exchange(
+    uintptr_t volatile* obj,
```
The default, not thread-safe implementation of this could be this:
if (*obj == expected) { *obj = desired; return true; }
return false;

antkmsft replied:
Thank you, this is so much better. I updated it.
